src/persist-client/src/operators/shard_source.rs

```diff
@@ -298,117 +222,277 @@ where
-                        let bytes_emitted = {
-                            let mut bytes_emitted = 0;
-                            for part_desc in std::mem::take(&mut batch_parts) {
-                                bytes_emitted += part_desc.encoded_size_bytes();
-                                // Give the part to a random worker. This isn't
-                                // round robin in an attempt to avoid skew issues:
-                                // if your parts alternate size large, small, then
-                                // you'll end up only using half of your workers.
-                                //
-                                // There's certainly some other things we could be
-                                // doing instead here, but this has seemed to work
-                                // okay so far. Continue to revisit as necessary.
-                                let worker_idx = usize::cast_from(Instant::now().hashed()) % num_workers;
-                                descs_session.give((worker_idx, part_desc.into_exchangeable_part()));
-                            }
-                            bytes_emitted
-                        };
-
-                        // Only track in-flight parts if flow control is enabled. Otherwise we
-                        // would leak memory, as tracked parts would never be drained.
-                        if flow_control_bytes.is_some() && bytes_emitted > 0 {
-                            inflight_parts.push((progress.clone(), bytes_emitted));
-                            inflight_bytes += bytes_emitted;
-                            trace!(
-                                "shard {} putting {} bytes inflight. total: {}. batch frontier {:?}",
-                                shard_id,
-                                bytes_emitted,
-                                inflight_bytes,
-                                progress,
-                            );
+        // Only one worker is responsible for distributing parts
+        if worker_index != chosen_worker {
+            trace!(
+                "We are not the chosen worker ({}), exiting...",
+                chosen_worker
+            );
+            return;
+        }
+
+        let create_subscribe = async {
+            let client = clients
+                .open(location)
+                .await
+                .expect("location should be valid");
+            let read = client
+                .open_leased_reader::<K, V, G::Timestamp, D>(
+                    shard_id,
+                    &format!("shard_source({})", name_owned),
+                    key_schema,
+                    val_schema,
+                )
+                .await
+                .expect("could not open persist shard");
+            let as_of = as_of.unwrap_or_else(|| read.since().clone());
+            (read.subscribe(as_of.clone()).await, as_of)
+        };
+        tokio::pin!(create_subscribe);
```

Comment:
probably a follow-up: could/should this use the new `pin!` from Rust 1.68? cc @guswynn

Reply:
updated!